grading/TestCase.cpp

```diff
@@ -188,8 +189,30 @@ bool openStudentFile(const TestCase &tc, const nlohmann::json &j, std::string &s
 bool openExpectedFile(const TestCase &tc, const nlohmann::json &j, std::string &expected_file_contents,
                       std::vector<std::pair<TEST_RESULTS_MESSAGE_TYPE, std::string> > &messages) {
 
-  std::string filename = j.value("expected_file","");
-  filename = getOutputContainingFolderPath(tc, filename) + filename;
+  std::cout << "opening expected file..." << std::endl;
+  std::string expected_file = j.value("expected_file", "");
+  std::string expected_string = j.value("expected_string", "");
+  std::string actual_file = j.value("actual_file", "");
+
+  std::cout << "expected_file: " << expected_file << std::endl;
+  std::cout << "expected_string: " << expected_string << std::endl;
+  std::cout << "actual_file: " << actual_file << std::endl;
+  assert(!(expected_file != "" && expected_string != ""));
+
+  std::string filename;
+  if (expected_file == "") {
+    //if the expected file doesn't exist, write expected_string to it (but where?)
+    //use actual_file name to get the name of the expected string output file
+    //ex: actual_file = math_1.txt, expected = AUTO_GENERATED_math_1.txt
+    expected_file = "AUTO_GENERATED_" + actual_file.substr(actual_file.find_last_of('/') + 1);
+    filename = "test_output/" + expected_file;
+    std::ofstream expected_file_stream(filename);
+    expected_file_stream << expected_string;
+    expected_file_stream.close();
+  } else {
+    filename = getOutputContainingFolderPath(tc, expected_file) + expected_file;
+  }
+  
   if (filename == "") {
     messages.push_back(std::make_pair(MESSAGE_FAILURE,"ERROR!  EXPECTED FILENAME MISSING"));
```

Comment:
If both `expected_string` and `expected_file` are missing, the `filename` will be `test_output/AUTOGENERATED_`; so the `if` statement below cannot catch the error.